README.md
# OS_GroupA_Project_02


History.txt
17,AMY HART,DEPARTMENT HEAD V,268604.57,0,16115.86,FT,0.45,2,160,3,0,1
17,AMY HART,DEPARTMENT HEAD V,268604.57,0,16115.86,FT,0.45,2,160,3,0,1
17,AMY HART,DEPARTMENT HEAD V,268604.57,0,16115.86,FT,0.45,2,160,3,0,1
17,AMY HART,DEPARTMENT HEAD V,268604.57,0,16115.86,FT,0.45,2,160,3,0,1



1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0,400184.25,PT,0.38,2,157,3,0,1




Server.c
/**
 * Author: Dillon Evans
 * Email: <dillon.e.evans@okstate.edu>
 * Date: March 8th, 2020
 * Program Description: This file creates a socket server that accepts queries from a client.
 * It searches the ID file for the name, then creates two threads to aggregate data
 * efficiently. The threads are then joined and the results are returned to the client.
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <pthread.h>
#include <ctype.h>
#include "defs.h"
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>

typedef struct{
    int desiredID;
    char* fileName;
}
fileParam;

bool equalsIgnoreCase(char* str1, char* str2){
    if (strlen(str1) != strlen(str2)){
        return FALSE;
    }
    else{
        for (int i = 0; i < strlen(str1); i++){
            if (tolower(str1[i]) != tolower(str2[i])){
                return FALSE;
            }
        }
    }
    return TRUE;
}
void *readFile(void *arg){
    //Implicitly cast the function argument to the struct containing arguments.
    fileParam *p = arg;
    int desiredID = p->desiredID;
    char *fileName = p->fileName;
    FILE* infile = fopen(fileName,"r");
    char buffer[1024] = {0};
    char name[256] = {0};
    bool foundMatch = FALSE;
    int id = 0;
    int count = -1;
    char *returnString = malloc(sizeof(char) * 1024);
    while (fgets(buffer, 1024, infile)){
        if (count >= 0){
            sscanf(buffer, "%d,%[^\r\n]", &id, returnString);
            if (id == desiredID){
                //printf("%s", buffer);
                break;
            }
        }
        count++;
    }
    fclose(infile);
    //sprintf(returnString, "%s", buffer);
    return returnString;
}

int main(){

    int server_socket;
    server_socket = socket(AF_INET,SOCK_STREAM, 0);
    struct sockaddr_in server_address;
    server_address.sin_family = AF_INET;
    server_address.sin_port = htons(9002);
    server_address.sin_addr.s_addr = INADDR_ANY;
    bind(server_socket, (struct sockaddr*)&server_address, sizeof(server_address));

    listen(server_socket, 1);

    int client_socket;
    client_socket = accept(server_socket, NULL, NULL);

    while(1){

        char query[256] = {0};
        recv(client_socket, query, sizeof(query), 0);
        printf("%s\n", query);
        FILE* infile = fopen("ID_name.txt","r");
        char buffer[1024] = {0};
        char name[256] = {0};
        bool foundMatch = FALSE;
        int id = 0;
        int count = -1;
        while (fgets(buffer, 1024, infile)){
            if (count >= 0){
                sscanf(buffer, "%d,%[^\r\n]", &id, name);
                if (equalsIgnoreCase(query, name)==TRUE){
                    printf("MATCH FOUND\n");
                    foundMatch = TRUE;
                    break;
                }
            }
            count++;
        }
        fclose(infile);

        if(!foundMatch){
            printf("The Employee %s Was Not Found.\n", query);
            send(client_socket, INVALID_QUERY, sizeof(INVALID_QUERY), 0);
        }else{
            pthread_t Salaries, Satisfaction;
            fileParam t1 = {id, "Salaries.txt"}, t2 = {id, "SatisfactionLevel.txt"};
            void *result1, *result2;
            pthread_create(&Salaries, NULL, readFile, &t1);
            pthread_create(&Satisfaction, NULL, readFile, &t2);

            pthread_join(Salaries, &result1);
            pthread_join(Satisfaction, &result2);

            char *str1 = result1, *str2 = result2;
            printf("%s\n", str1);
            printf("%s\n", str2);
            char returnString[1024] = {0};
            sprintf(returnString, "%d,%s,", id, name);
            strcat(returnString, str1);
            strcat(returnString,",");
            strcat(returnString, str2);

            free(str1);
            free(str2);
            printf("%s", returnString);
            printf("Return String: %s\n", returnString);
            send(client_socket, returnString, sizeof(returnString), 0);
        }
        //printf("%s", readFile(id, "Salaries.txt"));
        //printf("%s", readFile(id, "SatisfactionLevel.txt"));
    }

}


assist.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <unistd.h>
#include <wait.h>
#include "Assistant.h"
#include <stdbool.h>
#define FILEPATH "History.txt"

int LinePosition = 1;
char path[20] = "/dev/pts/";
bool isFirst = true;

void assistant(char *outgoingBuffer){
	//Storage for stuff from manager
	char *name[30];
	char *title[30];
	char *status[30];
	int fd;
	char pipe[] = ".pipe";

	//History file variables
  int LineCount = 0;
  char storeC;
  char outgoingBuffer[256];

	//Variables
  int count = 0;
  char incomingBuffer[256];

	//Setting up socket connection
	int network_socket = socket(AF_INET,SOCK_STREAM, 0);
	struct sockaddr_in server_address;
	server_address.sin_family = AF_INET;
	server_address.sin_port = htons(9002);
	server_address.sin_addr.s_addr = INADDR_ANY;

	/*Connect to the server and verify that the connection succeeded*/
	int connection_status = connect(network_socket, (struct sockaddr*)&server_address, sizeof(server_address));
    if(connection_status == -1){
        perror("There was an error establishing a connection");
        exit(EXIT_FAILURE);
    }

	/* create the FIFO (named pipe) */
	mkfifo(pipe, 0666);

	// Open pipe as read only for the assistant
	if((fd = open(pipe, O_RDONLY | O_CREAT)) < 0) perror("Pipe creation failure");

	while(TRUE){
		//reading info from pipe
		if(read(fd, &outgoingBuffer, sizeof(outgoingBuffer) + 1) < 0) perror("Read failure");
		sscanf(outgoingBuffer, "%[^,]%*c%[^,]%*c%[^\n]%*c", &name, title, status);


    	printf("%s\n", outgoingBuffer);

    // check number of lines written in file
    FILE *ftemp;
    FILE *fp = fopen("History.txt", "r");

    if(fp == NULL){
	    printf("unable to open file\n");
	    exit(EXIT_FAILURE);
    }

    while((storeC = fgetc(fp)) != EOF){
	    if(storeC == '\n'){
		    LineCount++;
	    }
    }

		// Go looking for the employee here
		send(network_socket, outgoingBuffer, 256, 0);
   	recv(network_socket, incomingBuffer, sizeof(incomingBuffer), 0);
	 	if(strcmp(incomingBuffer, "INVALID")==0){
			//if file has 10 lines use position to overwrite records
		}else if(LineCount == 10){
			//program uses temp.txt to store records from History.txt
			fp = fopen("History.txt", "r");
			ftemp = fopen("temp.txt", "w");

			while((fgets(lineBuffer, 256, fp)) != NULL){
				count++;

				if(count == LinePosition){
					fprintf(ftemp, "%s\n", incomingBuffer);
				}else{
					fprintf(ftemp, "%s", lineBuffer);
				}
			}

			fclose(fp);
			fclose(ftemp);

			//replace History.txt with the temp.txt
			remove("History.txt");
			rename("temp.txt", "History.txt");
			LinePosition++;

			if(LinePosition == 11){
				LinePosition = 1;
			}
		}else{
			fp = fopen("History.txt", "a");
			fprintf(fp, "%s\n", incomingBuffer);
			fclose(fp);
		}

		termPrinter(incomingBuffer);
	}



		close(network_socket);
    }
}

//function that prints to a new terminal
//requires tty ID for printing
void termPrinter(char *results){
	if(isFirst){
		printf("please type terminal tty ID number(type tty into the terminal you wish to print on and enter the last digit shown)\n");
		char id[2];
		scanf("%[^\n]%*c", id);
		strcat(path, id);
		isFirst = false;
	}

	FILE *fterm = fopen(path, "wr");

	if(fterm != NULL){
		fprintf(fterm, "%s\n", results);
	}else{
		perror("tty terminal not found");
		exit(EXIT_FAILURE);
	}

	fclose(fterm);
}


manager.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>	// for fork
#include <sys/types.h>	// for pid_t
#include <sys/stat.h>  //for mkfifo
#include <stdlib.h>	// for exit(1)
#include <string.h>	// for strlen
#include <sys/wait.h>	// for wait
#include <fcntl.h> //for file stuff
#include "defs.h" //our definitions

void manager() {
  char temp[30];
  char toSend[93];
  int fd;
  char pipe[] = ".pipe";

  /* create the FIFO (named pipe) */
  mkfifo(pipe, 0666);

  // Open pipe as write only for the manager
  if((fd = open(pipe, O_WRONLY | O_CREAT)) < 0) perror("Pipe failure");

  while(TRUE){ //Get user info and put in a single character array
    printf("Hello, What employee would you like information on?\n");
    scanf("%[^\n]%*c", temp);
    strcat(toSend,  temp);
    strcat(toSend, ",");
    memset(temp, 0, sizeof(temp));

    printf("What is their job title?\n", temp);
    scanf("%[^\n]%*c", temp);
    strcat(toSend,  temp);
    strcat(toSend, ",");
    memset(temp, 0, sizeof(temp));

    printf("What is their job status?\n", temp);
    scanf("%[^\n]%*c", temp);
    strcat(toSend,  temp);
    memset(temp, 0, sizeof(temp));

    printf("You are looking for: %s\n", toSend);
    printf("The assistant will report back to you shortly in their terminal\n\n");

    //writing info to pipe for the assistant
    if(write(fd, toSend, sizeof(toSend) + 1) < 0) perror("Write Failure");
    memset(toSend, 0, sizeof(toSend));
    sleep(5);
  }

  // Close plpe
  close(fd);

  /* remove the FIFO */
  unlink(pipe);
}


historySearch.c
//GROUP A
//Name: Okonkwo, Thomas Chukwuma
//CWID: A20308887
//Email: thomas.okonkwo@okstate.edu

#include "defs.h"
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include "Assistant.h"
bool equalsIgnoreCase(char* str1, char* str2){

    //compares the length of two strings passed
    if (strlen(str1) != strlen(str2)){
        return FALSE;
    }else{
        //loops through the string and convert to lower case
        //so that we can compare each letter of the strings
        for (int i = 0; i < strlen(str1); i++){
            if (tolower(str1[i]) != tolower(str2[i])){
                return FALSE;
            }
        }
    }
    return TRUE;
}

//this function searches the text file to see if a string is present in it
bool historySearch(char *query){
    //we read from the history file
    FILE* infile = fopen("History.txt","r");

    //created string variables/placeholders for the split of the query string
    //because we will have to search with the query string to see if the content
    //of the query is present in the history file. SO it's best practice to split
    //the query string into different string variables so we can search the history
    //file easily with these new variables.
    char nameQuery[256] = {0}, jobQuery[256] = {0}, statusQuery[256] = {0};
    char job[256] = {0}, name[256] = {0}, status[256] = {0};
    char buffer[1024] = {0};
    bool foundMatch = FALSE;
    float id = 0;
    int count = 0;

    //since in the original file, it is comma seperated, I decided to format
    //the string query into 3 different string variable names. This is so we
    //can search the history file with each of these string variables instead
    //of using the query string.
    sscanf(query, "%[^,],%[^,],%[^,]", nameQuery, jobQuery, statusQuery);

    //created a temp buffer that copies the content of the history file read
    while (fgets(buffer, 1024, infile)){
        //so we split the buffer content as well in to name, job and status so we can use
        //them to compare with the query string variable(split in 3 as shown above).
        sscanf(buffer, "%*d,%[^,],\"%[^\"]\",%*f,%*f,%*f,%[^,],%*[^\n\r]",name,job,status);

        //compares the job string to see if it's empty and format the main string into
        //name, job and status.
        if (strcmp(job, "") == 0){
            sscanf(buffer, "%*d,%[^,],%[^,],%*f,%*f,%*f,%[^,],%*[^\n\r]",name,job,status);

        }
        //compares and checks for the case sensitivity and ignores case sensitivity
        if (equalsIgnoreCase(name, nameQuery) && equalsIgnoreCase(job, jobQuery) && equalsIgnoreCase(status, statusQuery)){
	    termPrinter("Record found in File");
	    termPrinter(buffer); //passed to the termPrinter function
//	    printf("Query: %s\n", query);
//	    printf("Name: %s, Job: %s, Status: %s.\n", name, job, status);
            return TRUE;
        }
        strcpy(job, "");
    }
    //passed to the assistant function
    Assistant(nameQuery);
    fclose(infile);  //closed the file
    return FALSE;
}


assistant.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <netinet/in.h>
#include <unistd.h>
#include <wait.h>
#include "defs.h"
#include <stdbool.h>
#include <string.h>
#include <ctype.h>
#include <fcntl.h>
#define FILEPATH "History.txt"

//Group A
//Nikul Patel - worked on what to do after query was sent to server, and termPrinter function
//nikul.patel@okstate.edu
//CWID: A20066980

//void termPrinter(char *results);
bool equalsIgnoreCase(char* str1, char* str2);
FILE *fterm;
int LinePosition = 1;
char path[20] = "/dev/pts/";
bool isFirst = TRUE;


void assistant(){
	//Pipe variables
	int fd;
	char pipe[] = ".pipe";

	//History file variables
	FILE* ftemp;
    FILE* infile = fopen("History.txt","r");
    char nameQuery[256] = {0}, jobQuery[256] = {0}, statusQuery[256] = {0};
    char job[256] = {0}, name[256] = {0}, status[256] = {0};
    char buffer[1024] = {0}, query[1024] = {0};
    bool foundMatch = FALSE;
    float id = 0;
    int count = 0;
	bool history = FALSE;
    char incomingBuffer[256];

	/* create the FIFO (named pipe) */
	mkfifo(pipe, 0666);

	// Open pipe as read only for the assistant
	if((fd = open(pipe, O_RDONLY | O_CREAT)) < 0) perror("Pipe creation failure");

	// Socket creation
    int network_socket = socket(AF_INET,SOCK_STREAM, 0);
    struct sockaddr_in server_address;
    server_address.sin_family = AF_INET;
    server_address.sin_port = htons(9002);
    server_address.sin_addr.s_addr = INADDR_ANY;

	//Connect to the server and verify that the connection succeeded
    int connection_status = connect(network_socket, (struct sockaddr*)&server_address, sizeof(server_address));
    if (connection_status == -1){
        perror("There was an error establishing a connection");
        exit(EXIT_FAILURE);
    }

	// check number of lines written in History.txt
    int LineCount = 0;
    char storeC;
    char lineBuffer[256];
    if(infile == NULL){
	    printf("unable to open file\n");
	    exit(EXIT_FAILURE);
    }

    while((storeC = fgetc(infile)) != EOF){
	    if(storeC == '\n'){
		    LineCount++;
	    }
    }


	while(TRUE){
		//Get query from manager through the pipe
		if(read(fd, &query, sizeof(query) + 1) < 0) perror("Read failure");
		sscanf(query, "%[^,]%*c%[^,]%*c%[^\n]%*c", nameQuery, jobQuery, statusQuery);
		//Check for info in History.txt first

		if(!historySearch(query)){
			//Send name to server to see if it contains the employee
			send(network_socket, nameQuery, 256, 0);
			recv(network_socket, incomingBuffer, sizeof(incomingBuffer), 0);
			//Wait for a non empty message to be received.
		    while (strcmp(incomingBuffer, "") == 0) {
				recv(network_socket, incomingBuffer, sizeof(incomingBuffer), 0);
			}
			if(strcmp(incomingBuffer, "INVALID")==0){}
			//if file has 10 lines use position to overwrite records
			else if(LineCount == 10){
				//program uses temp.txt to store records from History.txt
				infile = fopen("History.txt", "r");
				ftemp = fopen("temp.txt", "w");
				while((fgets(lineBuffer, 256, infile)) != NULL){
				count++;
					if(count == LinePosition){
						fprintf(ftemp, "%s\n", incomingBuffer);
					}else{
						fprintf(ftemp, "%s", lineBuffer);
					}
				}

				fclose(infile);
				fclose(ftemp);
				//replace History.txt with the temp.txt
				remove("History.txt");
				rename("temp.txt", "History.txt");
				LinePosition++;
				if(LinePosition == 11){
					LinePosition = 1;
				}
			}else{
				infile = fopen("History.txt", "a");
				fprintf(infile, "%s\n", incomingBuffer);
				fclose(infile);
			}
			termPrinter(incomingBuffer);
		}
	}
	close(network_socket);
}


//function that prints to a new terminal
//requires tty ID for printing
void termPrinter(char *results){
	if(isFirst){
		printf("please type terminal tty ID number(type tty into the terminal you wish to print on and enter the last digit shown)\n");
		char id[2];
		scanf("%[^\n]%*c", id);
		strcat(path, id);
		printf("%s\n", path);
		isFirst = FALSE;
	}

	fterm = fopen(path, "wr");

	if(fterm != NULL){
		fprintf(fterm, "%s\n", results);
	}else{
		perror("tty terminal not found");
		exit(EXIT_FAILURE);
	}

	fclose(fterm);
}
